misc/python/materialize/mzcompose/composition.py

```diff
@@ -654,15 +655,92 @@ def up(
             self.compose = old_compose  # type: ignore
             self._write_compose()
 
-    def down(self, destroy_volumes: bool = True, remove_orphans: bool = True) -> None:
+    def validate_sources_sinks_clusters(self) -> str | None:
+        """Validate that all sources, sinks & clusters are in a good state"""
+
+        results = self.sql_query(
+            """
+            SELECT name, status, error, details
+            FROM mz_internal.mz_source_statuses
+            WHERE NOT(
+                status = 'running' OR
+                (type = 'progress' AND status = 'created') OR
+                (type = 'subsource' AND status = 'starting')
+            )
+            """
+        )
+        for (name, status, error, details) in results:
+            # TODO(def-) Remove when #21980 is fixed
+            if status == "starting":
+                continue
+
+            return f"Source {name} is expected to be running/created, but is {status}, error: {error}, details: {details}"
+
+        results = self.sql_query(
+            """
+            SELECT name, status, error, details
+            FROM mz_internal.mz_sink_statuses
+            WHERE status NOT IN ('running', 'dropped')
+            """
+        )
+        for (name, status, error, details) in results:
+            return f"Sink {name} is expected to be running/dropped, but is {status}, error: {error}, details: {details}"
+
+        results = self.sql_query(
+            """
+            SELECT mz_clusters.name, mz_cluster_replicas.name, status, reason
+            FROM mz_internal.mz_cluster_replica_statuses
+            JOIN mz_cluster_replicas
+            ON mz_internal.mz_cluster_replica_statuses.replica_id = mz_cluster_replicas.id
+            JOIN mz_clusters ON mz_cluster_replicas.cluster_id = mz_clusters.id
+            WHERE status NOT IN ('ready', 'not-ready')
+            """
+        )
+        for (cluster_name, replica_name, status, reason) in results:
+            return f"Cluster replica {cluster_name}.{replica_name} is expected to be ready/not-ready, but is {status}, reason: {reason}"
+
+        return None
+
+    def sanity_restart_mz(self) -> None:
+        """Restart Materialized if it is part of the composition to find
+        problems with persisted objects, functions as a sanity check."""
+        if "materialized" in self.compose["services"]:
+            self.kill("materialized")
+            # TODO(def-): Better way to detect when kill has finished
+            time.sleep(3)
+            self.up("materialized")
+            self.sql("SELECT 1")
```

Comment:
Should we make sure that this succeeds? Or is this done implicitly because an error will be raised? Or do we rely on the following `validate_sources_sinks_clusters()` logic?

Reply:
If the query doesn't  fail, that is already enough, the validate function is more interesting anyway.